### PR TITLE
Predictor step improvements

### PR DESCRIPTION
**Pull request by stefannica**
1. list all input parameters explicitly as environment variables
in the Dockerfiles

2. add `app_name` parameter: implements an `app_name` input
parameter that can be set for all predictor steps that determines
the name of the FuseML application as well as all k8s resources
created for it. This allows users to create multiple applications
(i.e. prediction services) within a single workflow without resulting
in name clashing. It also allows users to control the same application
from multiple workflows. Setting the `app_name` input parameter
explicitly also removes the need to mount a codeset into the predictor
step.

3. add `verbose` parameter that controls verbosity

4. fix checking of existing folders in minio bucket

### CODE REVIEW

File: images/inference-services/kfserving/Dockerfile [hunk 1]
```diff
@@ -30,4 +30,16 @@ RUN apk add --update gettext \
 COPY template.sh  /root/template.sh
 COPY run.sh /usr/local/bin/run
 
+ENV FUSEML_ENV_WORKFLOW_NAMESPACE fuseml-workloads
+ENV FUSEML_ENV_WORKFLOW_NAME ""
+ENV FUSEML_APP_NAME ""
+ENV FUSEML_VERBOSE false
```
Comment (jsuchome) [inline images/inference-services/kfserving/Dockerfile:36]: +1 !

File: images/inference-services/kfserving/template.sh [hunk 2]
```diff
@@ -14,22 +14,20 @@ stringData:
...
+  - name: "${APP_NAME}-storage"
 ---
 apiVersion: "serving.kubeflow.org/v1beta1"
 kind: "InferenceService"
 metadata:
-  name: "${isvc}"
+  name: "${APP_NAME}"
   labels:
-    fuseml/app-name: "${PROJECT}"
-    fuseml/org: "${ORG}"
```
Comment (jsuchome) [inline images/inference-services/kfserving/template.sh:27]: isn't it shame to lose these labels?

File: images/inference-services/ovms/run.sh [hunk 3]
```diff
@@ -60,7 +64,7 @@ resources: ${RESOURCES}
 target_device: "${FUSEML_TARGET_DEVICE}"
 EOF
 
-cat /opt/openvino/templates/values.yaml
+$FUSEML_VERBOSE && cat /opt/openvino/templates/values.yaml
```
Comment (jsuchome) [inline images/inference-services/ovms/run.sh:67]: don't you want to add this to kfserving and seldon too?

File: images/inference-services/seldon-core/run.sh [hunk 4]
```diff
@@ -69,14 +79,14 @@ until [[ -n "$(kubectl get deploy -l seldon-deployment-id=${sd} -n ${FUSEML_ENV_
...
+kubectl rollout status deploy/$(kubectl get deploy -l seldon-deployment-id=${APP_NAME} -n ${FUSEML_ENV_WORKFLOW_NAMESPACE} -o jsonpath='{.items[0].metadata.name}') -n ${FUSEML_ENV_WORKFLOW_NAMESPACE}
 
 prediction_url="http://${ISTIO_HOST}/seldon/${FUSEML_ENV_WORKFLOW_NAMESPACE}/${prediction_url_path}"
 
 printf "${prediction_url}" > /tekton/results/${TASK_RESULT}
 
 # Now, register the new application within fuseml
 
-resources="{\"kind\": \"Secret\", \"name\": \"${ORG}-${PROJECT}-init-container-secret\"}, {\"kind\": \"ServiceAccount\", \"name\": \"${ORG}-${PROJECT}-seldon\"}, {\"kind\": \"SeldonDeployment\", \"name\": \"${ORG}-${PROJECT}\"}"
+resources="{\"kind\": \"Secret\", \"name\": \"${APP_NAME}-init-container-secret\"}, {\"kind\": \"ServiceAccount\", \"name\": \"${APP_NAME}-seldon\"}, {\"kind\": \"SeldonDeployment\", \"name\": \"${APP_NAME}\"}"
```
Comment (jsuchome) [inline images/inference-services/seldon-core/run.sh:90]: oh, I completely missed these before, thanks!

File: images/inference-services/kfserving/run.sh [hunk 5]
```diff
@@ -34,19 +41,23 @@ if [ "${PREDICTOR}" = "auto" ]; then
...
-export isvc="${ORG}-${PROJECT}-${FUSEML_ENV_WORKFLOW_NAME}"
 case $PREDICTOR in
     # kfserving expects the tensorflow model to be under a numbered directory,
-    # however mlflow saves the model under 'tfmodel', so if there is no directory
+    # however mlflow saves the model under 'tfmodel' or 'data/model', so if there is no directory
     # named '1', create it and copy the tensorflow model to it.
-    tensorflow)
-        if [ "$(mc find ${model_bucket} --name 1)" = "" ]; then
-            mc cp -r "${model_bucket}"/tfmodel/ "${model_bucket}"/1
+    keras|tensorflow)
```
Comment (flaviodsr) [inline images/inference-services/kfserving/run.sh:48]: Why add `keras` as a predictor? AFAIK there is no `keras` kind of predictor that is supported by kfserving.

**Review by flaviodsr — CHANGES_REQUESTED** [review]
Overall the changes looks very nice. Just a note about the `keras` predictor.

Also, I am not convinced about the `app_name` it seems like a good idea however without proper checks, or some random suffix/prefix, it allows one to freely overwrite other applications. For example, if I have a codeset named `cifar100` under project `A` while other user has `cifar100` under project `B` there is a huge chance that both will choose `cifar100` as `app_name` hence both will be creating the same inference service, one overwriting the other.

File: images/inference-services/kfserving/run.sh (hunk 5 above)
Comment (stefannica) [reply images/inference-services/kfserving/run.sh:45]: You're right, there isn't. But MLFlow does encode this differently in the generated `MLmodel` file when it comes from Keras, and we need to deal with it here. This is only useful when the predictor parameter is set to `auto` and the model is a Keras/MLFlow model. A few lines below, I change it back to `tensorflow`.

Comment (flaviodsr) [reply images/inference-services/kfserving/run.sh:45]: But if the predictor parameter is set to `auto` `$PREDICTOR` will never be `keras`, see: https://github.com/fuseml/extensions/blob/58194eb24ceb30ea7a16173443c789de7caf67aa/images/inference-services/kfserving/run.sh#L36

File: images/inference-services/kfserving/Dockerfile (hunk 1 above)
Comment (stefannica) [reply images/inference-services/kfserving/Dockerfile:36]: thanks :)

File: images/inference-services/kfserving/template.sh (hunk 2 above)
Comment (stefannica) [reply images/inference-services/kfserving/template.sh:27]: ok, I revived them

File: images/inference-services/ovms/run.sh (hunk 3 above)
Comment (stefannica) [reply images/inference-services/ovms/run.sh:67]: done

File: images/inference-services/kfserving/run.sh (hunk 5 above)
Comment (stefannica) [reply images/inference-services/kfserving/run.sh]: Ok, so based on our conversation, using `keras` as a predictor value sends the wrong idea. I'll think of a different way to fix this. For now, I'm removing the keras changes from this PR.

**Review by jsuchome — APPROVED** [review]
thanks

**Review by flaviodsr — APPROVED** [review]
